File: project/plotting/plot.cpp
#include "BSpline.h"
#include "CurveFitting.h"
#include "PPSpline.h"
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

template <typename T>
void savePlottingData(const T& f, const std::string& filename, size_t numSample = 300) {
    std::ofstream outFile(filename);
    if (!outFile) {
        std::cerr << "Error: Could not open file " << filename << " for writing!" << std::endl;
        return;
    }

    f.plot(outFile, numSample);
    outFile.close();
    std::cout << "Saved spline points to " << filename << std::endl;
}

class C1 : public Curve {
  public:
    C1() : Curve(0, 2 * Pi) {}

    Point operator()(double t) const override {
        return {std::cos(t), std::sin(t)};
    }
};

int main() {

    std::string outputDir = "./plotting/data/";

    C1 c1;
    CurveFitting cf(c1, 0.1);
    savePlottingData(cf, outputDir + "curve_fitting.txt");

    return 0;
}

File: project/plotting/plot.py
import os
import matplotlib.pyplot as plt

def plot_spline(file_path, output_path):
    # Load data
    x, y = [], []
    with open(file_path, 'r') as f:
        for line in f:
            parts = line.strip().split()
            if len(parts) == 2:
                x.append(float(parts[0]))
                y.append(float(parts[1]))

    # Plot
    plt.figure(figsize=(8, 6))
    plt.plot(x, y, label="Spline Curve")
    plt.scatter(x, y, color="red", s=10, label="Points")
    plt.title(f"Spline Curve: {os.path.basename(file_path)}")
    plt.xlabel("X")
    plt.ylabel("Y")
    plt.legend()
    plt.grid(True)
    plt.savefig(output_path)
    plt.close()
    print(f"Saved plot to {output_path}")

def process_all_splines(input_dir, output_dir):
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    for file_name in os.listdir(input_dir):
        if file_name.endswith(".txt"):
            file_path = os.path.join(input_dir, file_name)
            output_path = os.path.join(output_dir, file_name.replace(".txt", ".png"))
            plot_spline(file_path, output_path)

if __name__ == "__main__":
    input_directory = "./plotting/data/"
    output_directory = "./plotting/figures/"
    process_all_splines(input_directory, output_directory)
